### PR TITLE
[Enhance] Enhance feature extraction function.

### PR DESCRIPTION
**Pull request by mzr1996**
## Motivation

Feature extraction is an important function of the backbone. We already have the `extract_feat` method in `ImageClassifier`, but it can only extract the feature map after neck.

## Modification
Here we split the whole network into six parts and enable our API to get all of their outputs.
流程图

#### ① backbone output
```python
import torch
from mmcv import Config
from mmcls.models import build_classifier

cfg = Config.fromfile('configs/resnet/resnet18_8xb32_in1k.py').model
cfg.backbone.out_indices = (0, 1, 2, 3)  # Output multi-scale feature maps
model = build_classifier(cfg)

outs = model.extract_feat(torch.rand(1, 3, 224, 224), stage='backbone')
for out in outs:
    print(out.shape)
# torch.Size([1, 64, 56, 56])
# torch.Size([1, 128, 28, 28])
# torch.Size([1, 256, 14, 14])
# torch.Size([1, 512, 7, 7])
```
#### ② neck output
The neck is usually `GlobalAveragePooling`, and some networks don't have a neck.
```python
import torch
from mmcv import Config
from mmcls.models import build_classifier

cfg = Config.fromfile('configs/resnet/resnet18_8xb32_in1k.py').model
cfg.backbone.out_indices = (0, 1, 2, 3)  # Output multi-scale feature maps
model = build_classifier(cfg)

outs = model.extract_feat(torch.rand(1, 3, 224, 224), stage='neck')
for out in outs:
    print(out.shape)
# torch.Size([1, 64])
# torch.Size([1, 128])
# torch.Size([1, 256])
# torch.Size([1, 512])
```
#### ③ Pre-logits output (without the final linear classifier head)
Some heads have not only a single linear layer classifier but also some other processing, like `VisionTransformerClsHead` and `StackedLinearClsHead`. Now we can extract features before the final linear classifier.
```python
import torch
from mmcv import Config
from mmcls.models import build_classifier

cfg = Config.fromfile('configs/vision_transformer/vit-base-p16_pt-64xb64_in1k-224.py').model
model = build_classifier(cfg)

out = model.extract_feat(torch.rand(1, 3, 224, 224), stage='pre_logits')
print(out.shape)  # The hidden dims in head is 3072
# torch.Size([1, 3072])
```
#### ④ Head linear output (without softmax)
```python
import torch
from mmcv import Config
from mmcls.models import build_classifier

cfg = Config.fromfile('configs/vision_transformer/vit-base-p16_pt-64xb64_in1k-224.py').model
model = build_classifier(cfg)

out = model.simple_test(torch.rand(1, 3, 224, 224), softmax=False, post_process=False)
print(out.shape)
# torch.Size([1, 1000])
```
#### ⑤ Softmax output (without post-processing)
```python
import torch
from mmcv import Config
from mmcls.models import build_classifier

cfg = Config.fromfile('configs/vision_transformer/vit-base-p16_pt-64xb64_in1k-224.py').model
model = build_classifier(cfg)

out = model.simple_test(torch.rand(1, 3, 224, 224), post_process=False)
print(out.shape)
# torch.Size([1, 1000])
```
In multi-label tasks, the `softmax` is changed to `sigmoid`.
#### ⑥ Post-processing output
In post-processing, we will convert the tensor output into a list.
```python
import torch
from mmcv import Config
from mmcls.models import build_classifier

cfg = Config.fromfile('configs/vision_transformer/vit-base-p16_pt-64xb64_in1k-224.py').model
model = build_classifier(cfg)

out = model.simple_test(torch.rand(1, 3, 224, 224))
print(f'{type(out)} ({len(out)}, {len(out[0])})')
# <class 'list'> (1, 1000)
```
The post-processing doesn't depend on softmax, you can also post-process the logits output.
```python
out = model.simple_test(torch.rand(1, 3, 224, 224), softmax=False)
```

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by fcakyon** [comment]
@mzr1996 this is an awesome PR. Just what I needed! Great job. Cant wait to get it merged @Ezra-Yu :+1:

File: mmcls/models/classifiers/base.py [hunk 1]
```diff
@@ -35,13 +35,14 @@ def with_head(self):
         return hasattr(self, 'head') and self.head is not None
 
     @abstractmethod
-    def extract_feat(self, imgs):
+    def extract_feat(self, imgs, stage=None):
         pass
 
-    def extract_feats(self, imgs):
+    def extract_feats(self, imgs, stage=None):
```
Comment (Ezra-Yu) [inline mmcls/models/classifiers/base.py:41]: `imgs` could be a python `Sequeue`

**Review by Ezra-Yu — COMMENTED** [review]
1. remove se_cfg in  the docstring in  `StackedLinearClsHead`
2. add some docstring and for some functions and class

**Review by Ezra-Yu — APPROVED** [review]
LGTM.

**Comment by greattBILL** [comment]
Hello, are all the codes based on the default, why is there no information about the trained model and the relative data set? I want to use the model I have trained and the matching data set picture for neck output, can you provide a code demonstration. Thank you very much？

**Comment by greattBILL** [comment]
Hello, are all the codes based on the default, why is there no information about the trained model and the relative data set? I want to use the model I have trained(such as resnet101) and the matching dataset picture(imagenet) for neck output, can you provide a code demonstration. Thank you very much.

**Comment by mzr1996** [comment]
> Hello, are all the codes based on the default, why is there no information about the trained model and the relative data set? I want to use the model I have trained(such as resnet101) and the matching dataset picture(imagenet) for neck output, can you provide a code demonstration. Thank you very much.

These codes are just simple examples. If you want to use the real dataset, you also need to build the pipeline to do some pre-processing. Here is an example:
```python
import mmcv
from mmcv import Config
from mmcls.apis import init_model
from mmcls.datasets import build_dataset

cfg = Config.fromfile('configs/resnet/resnet18_8xb32_in1k.py')
cfg.model.backbone.out_indices = (0, 1, 2, 3)  # Output multi-scale feature maps
model = init_model(cfg, 'your_checkpoint_path', device='cpu')  # Use the API `init_model` to initialize and load checkpoint

dataset = build_dataset(cfg.data.test)  # Use test dataset and test pipeline
img = dataset[0]['img'][None, :]    # Get one input image
outs = model.extract_feat(img, stage='neck')
for out in outs:
    print(out.shape)
# torch.Size([1, 256])
# torch.Size([1, 512])
# torch.Size([1, 1024])
# torch.Size([1, 2048])
```

**Comment by jingmingtao** [comment]
import mmcv
from mmcv import Config
from mmcls.apis import init_model
from mmcls.datasets import build_dataset

cfg = Config.fromfile('configs/resnet/resnet18_8xb32_in1k.py')
cfg.model.backbone.out_indices = (0, 1, 2, 3)  # Output multi-scale feature maps
model = init_model(cfg, 'your_checkpoint_path', device='cpu')  # Use the API `init_model` to initialize and load checkpoint

dataset = build_dataset(cfg.data.test)  # Use test dataset and test pipeline
img = dataset[0]['img'][None, :]    # Get one input image
outs = model.extract_feat(img, stage='neck')
for out in outs:
    print(out.shape)
# torch.Size([1, 256])
# torch.Size([1, 512])
# torch.Size([1, 1024])
# torch.Size([1, 2048])
请教在新版的mmpretrain中如何完成该特征提取的功能，尝试了很久无法完成：
import mmengine
from mmengine import Config
from mmpretrain.apis import init_model
from mmpretrain.datasets import build_dataset

from mmpretrain.models import build_classifier
import torch

cfg = Config.fromfile('configs/resnet/resnet18_8xb32_in1k.py')
cfg.model.backbone.out_indices = (0, 1, 2, 3)  # Output multi-scale feature maps
model = init_model(cfg, 'your_checkpoint_path', device='cpu')  # Use the API `init_model` to initialize and load checkpoint

dataset = build_dataset(cfg.data.test)  # Use test dataset and test pipeline
img = dataset[0]['img'][None, :]    # Get one input image
outs = model.extract_feat(img, stage='neck')
for out in outs:
    print(out.shape)

**Comment by greattBILL** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by greattBILL** [comment]
这是来自QQ邮箱的假期自动回复邮件。
 
您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。
